File: network/src/main/java/pl/malopolska/smoksmog/DateTimeDeserializer.java
package pl.malopolska.smoksmog;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;

import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;

import java.lang.reflect.Type;
import java.util.TimeZone;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class DateTimeDeserializer implements JsonDeserializer<DateTime> {

    private static final Pattern DATE_TIME_PATTERN = Pattern.compile( "(\\d+)-(\\d+)-(\\d+)\\s+?(\\d+):(\\d+):(\\d+)" );

    @Override
    public DateTime deserialize( JsonElement json, Type typeOfT, JsonDeserializationContext context ) throws JsonParseException {

        Matcher matcher = DATE_TIME_PATTERN.matcher( json.getAsString() );

        if(!matcher.matches()){
            throw new JsonParseException( "Invalid date format" );
        }

        int year = Integer.parseInt( matcher.group( 1 ) );
        int month = Integer.parseInt( matcher.group( 2 ) );
        int day = Integer.parseInt( matcher.group( 3 ) );
        int hour = Integer.parseInt( matcher.group( 4 ) );
        int minute = Integer.parseInt( matcher.group( 5 ) );
        int second = Integer.parseInt( matcher.group( 6 ) );

        DateTime dateTime = DateTime.now( DateTimeZone.forTimeZone( TimeZone.getTimeZone( "Europe/Warsaw" ) ) )
                .withYear( year )
                .withMonthOfYear( month )
                .withDayOfMonth( day )
                .withHourOfDay( hour )
                .withMinuteOfHour( minute )
                .withSecondOfMinute( second );

        return dateTime;
    }
}

File: network/src/main/kotlin/pl/malopolska/smoksmog/DateTimeDeserializer.kt
package pl.malopolska.smoksmog

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import org.joda.time.DateTime
import org.joda.time.DateTimeZone
import java.lang.reflect.Type
import java.util.*

class DateTimeDeserializer : JsonDeserializer<DateTime> {

    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): DateTime {

        val input = json.asString

        val matcher = "(\\d+)-(\\d+)-(\\d+)\\s+?(\\d+):(\\d+):(\\d+)".toPattern().matcher(input)

        if (!matcher.matches()) {
            throw JsonParseException("Invalid date format")
        }

        val year = Integer.parseInt(matcher.group(1))
        val month = Integer.parseInt(matcher.group(2))
        val day = Integer.parseInt(matcher.group(3))
        val hour = Integer.parseInt(matcher.group(4))
        val minute = Integer.parseInt(matcher.group(5))
        val second = Integer.parseInt(matcher.group(6))

        val dateTime = DateTime.now(DateTimeZone.forTimeZone(TimeZone.getTimeZone("Europe/Warsaw")))
                .withYear(year)
                .withMonthOfYear(month)
                .withDayOfMonth(day)
                .withHourOfDay(hour)
                .withMinuteOfHour(minute)
                .withSecondOfMinute(second)

        return dateTime
    }
}
